### ISSUE TITLE
efl: Issues with Transparency/Alpha on elm_bg - 1.21 regression

### ISSUE THREAD

**Issue by Peter2121** [Obsidian-StudiosInc/ecrire#29]
...compiled with GIT version of EFL. Just white background in the working area, no text.

**Comment by wltjr** [comment]
Wonderful news, EFL seems to be a mess. They keep breaking the old/legacy API and other issues as they move to EO and also Wayland. Seeing discussions on developer list they do not have a clear path forward. Sadly elm code author rather work on documentation and other things than fixing bugs... Likely not much I can do here other than see if it is fixed with further EFL development before next release. Also why I stick to release tarballs vs git. I cant take the chance on random breakage that seems common in git. Probably the worst graphics libraries and API for *nix.

**Comment by haxworx** [comment]
Elm Code has never been released as a stable API.  It's probably broken because we're adding more features.

**Comment by haxworx** [comment]
Also as an application developer why do you encourage users to build against an ever changing EFL git??? They're using unstable release. Your accusations are false and unhelpful.

**Comment by Peter2121** [comment]
I think that the return to simple multi-line text widget would be nice, waiting for the stabilization of Elm Code. Personally, I just search for leafpad replacement...

**Comment by haxworx** [comment]
That sounds reasonable Peter2121

**Comment by wltjr** [comment]
@haxworx I did not recommend nor suggest ever that anyone should use anything live from git. If you read my comments I specifically say just the opposite...

"Also why I stick to release tarballs vs git"
https://github.com/Obsidian-StudiosInc/ecrire/issues/29#issuecomment-352800638

Regarding using elm_code, that was @ajwillia-ms suggestion that I switch from elm_entry to elm_code sometime back in IRC. Only way I even knew about elm_code. I very much regret that change...

@haxworx thanks for your hostility despite being completely wrong.... Now you know the facts...

**Comment by haxworx** [comment]
William I think you should strongly consider not using elm_code until it is deemed stable, or you're likely to come across these sorts of issues a lot.

**Comment by wltjr** [comment]
@haxworx you need to talk to @ajwillia-ms not me... It was recommended I switch to elm_code vs implement line numbers etc on my own... Thanks for missing the back story... I did not switch to elm_code on a whim. It was not my idea and I very much regret.

Given the amount of work I did to switch over I do not feel like switching back. Maybe you should be focusing on working and fixing bugs so things can move forward. Rather than a regression because elm_code is in such flux.

Just the same @ajwillia-ms should NOT be telling people to use elm_code. Just for you @haxworx to come along and say the complete opposite...

Maybe you all need to get your act together and stop causing these issues for others. I would have gone a completely different route had it not been for @ajwillia-ms. Like what they did with epad... Not to mention the python-elm-extensions. I already had other ideas for using third party libraries for syntax and other things.

Really ridiculous....

**Comment by haxworx** [comment]
If you want to stay with elm_code am more than happy to discuss changes just prior to "stable" release of EFL and a changed  elm_code. it's too much to work follow and try to stay on top of things if with an ever-changing development EFL.

**Comment by wltjr** [comment]
@haxworx you need to go speak with @ajwillia-ms and EFL get its act together. Stop harping on me for following what the elm_code author told me... You are coming in after the fact and missing the entire story...

**Comment by wltjr** [comment]
@haxworx I keep up with development in many places, Java 9... Please this is a EFL community issue. There are many messes created upstream. You all need to get your act together not me...

**Comment by haxworx** [comment]
We're quite happy working on EFL and elm_code but would not suggest using the elm_code API at the moment as it's in a state of change as we are improving it. Thanks.

**Comment by wltjr** [comment]
@haxworx how many times do I need to tell you that it was suggested I switch from elm_entry to elm_code? I did not do that without another recommending I do exactly that.... @ajwillia-ms

**Comment by haxworx** [comment]
Peter2121 I'd suggest looking at the original ecrire which does not use an unstable API.

**Comment by wltjr** [comment]
@haxworx do you have a problem? I removed all that code, I am not adding it back... I did not do that on a whim. How many times can I say I regret listening to @ajwillia-ms in the first place? I am hardly going to take your advice now... @haxworx  if your not fixing the issue, please do not suggest other routes. You clearly do not understand the time I put into this and I am not looking to repeat that. Because in a year you call cant get basic functionality stable....

@haxworx are you the author of elm_code or is @ajwillia-ms? Why are you both saying different things? Again you two need to go talk and form a consensus....

I can post full logs. This was first suggested in 4.2.2017.
#e/2017-04-02.log:14:50 < ajwillia-ms> wltjr: FYI if epad switched to the elm_code widget (beta) it would get the same syntax highlighting as edi :)

#e/2017-04-19.log:22:38 < wltjr> raster: I discussed it all with edi author/maintainer, and they mentioned the elm_code thing

#e/2017-04-19.log:22:44 < wltjr> raster: but edi maintainer is one who said use
elm_code, discussed it with them

**Comment by haxworx** [comment]
As I said, prior to a release am more than happy to discuss changes to behaviours.

**Comment by wltjr** [comment]
@haxworx do you understand beta vs alpha? Go talk to @ajwillia-ms and stop making noise here. Seriously haven't you anything better to do? I surely do and this is not addressing any issues....

**Comment by haxworx** [comment]
I'm sure you're perfectly capable of thinking for yourself. If you want help prior to a release there's no issue discussing changes to elm_code. So, if you stick to release tarballs like you do, then prior to a release can make time to introduce you to improvements or changes affecting your port.

**Comment by wltjr** [comment]
@haxworx you should stop doing EFL development. I haven't the time to go discussing changes with authors for all releases. That is called release notes... Major things should not change or break. 

If elm_code is what you say. It should not even be in EFL if it is not ready for others to use.... I have asked you politely several times. Please for the last time stop commenting unless you are fixing the issue from a technical perspective. Nothing you have said is helpful in any form.

@haxworx you are simply wasting time... please stop!

**Comment by haxworx** [comment]
I've some free time, am happy to be here offering help.

**Comment by wltjr** [comment]
I don't have free time, I am working on Java 9 stuff.... There are plenty of open bugs for you to spend your free time fixing.... I have asked politely to many times. Please stop commenting on this issue.

Unless  you have something of technical value and/or resolving issues. Please do not comment. You are not adding anything of value. Next I will just block which I rather not. But I have other work this is taking me from... I am VERY busy... I haven't free time almost ever so please for the last time. STOP wasting my time. Go fix YOUR elm_code.... Or make more API changes that break others applications...

**Comment by haxworx** [comment]
Thanks William. I'd ask you'd do the same and stop leaving abuse on our mailing list.

**Comment by wltjr** [comment]
@haxworx you can't handle the truth, your code sucks. If others call you out, you try to turn around and blame it on them. Rather than go fixing your code.... I can leave the mailing list. It surely will not make any of this any better.

Not that EFL or EFL devs care about quality, stability or other issues. Your lack of respect for others time and repeat comments. You have not used please once.... Rude...

**Comment by haxworx** [comment]
That would be great if you wouldn't mind.

**Comment by wltjr** [comment]
@Peter2121 sorry for all that noise. You can see what I am dealing with upstream now... Its a total mess...

**Comment by andydotxyz** [comment]
@wltjr I will ask this one last time: please stop abusing Enlightenment community developers - that is not an effective way to get better support.

@haxworx is one of our most active EFL app developers and is also a developer of elm_code. He has done lots of great work and does not deserve your abuse.

This issue is not reproducible upstream (check elementary_test) I can’t even think why you are seeing this issue as no other elm_code api users have seen it. 

Please untag upstream_bug unless you have more info for use to work with.

**Comment by wltjr** [comment]
@ajwillia-ms I am not reporting this issue, nor can I replicate as I do not run EFL from git... I did not ask @haxworx to come here. Nothing he said was of benefit. If you think telling someone to stop using elm_code is a solution to technical issues with elm_code. I fail to see how that makes elm_code any better.

@ajwillia-ms I also do not appreciate you recommending me to use something. That another developer turns around and says not to use. That I should revert all my coding efforts around elm_code for the past many months. Why did I even bother with it in the first place? Why did you mention elm_code to me in the first place? Why recommend to use?

The EFL and E community needs to think about the impact of their actions on others. If you do not like my tone. Then do not release crappy software. Or tell others to use such things that are not even alpha, nor should be in EFL. Stop trying to blame me for problems you all are creating.

You all did something in EFL that broke a EFL application. It is NOT application developers role to ensure that API developers are not breaking their applications. That is the role of a API developers to not mess up the API others are relying on. It shows a complete lack of consideration to others, not to mention inexperience...

I have Java GUIs close to 20 years old that still run fine. My old GTK code from 06 still runs fine. Those APIs are reliable and stable. Those developers do not require people who developed applications to check in with them for any API change. Most all provide some layer to not break legacy applications. Or give them a path. GTK 3 did not break GTK 2 stuff, nor did versions of QT. All that was done properly and applications were ported to the newer version. Go look to see when Java Swing will be removed for JavaFX. It will take years, even AWT remains for the entire life of Swing....

I have been packaging Netbeans 9 from source not using ant or their build system. I have used it since going back to at least 3.6, across many versions of Java, and now Java 9, with Netbeans 9. Only till recently did oracle do anything that broke other stuff. Though they provided advanced warning, and most that stuff was closed source, Sun Microsystems classes etc. That others likely should never have used. Thus oracle phasing it out, and it being a problem. For the VAST majority of applications. Java 9 introduces no breakage. I have applications developed in Java 1.4, that run fine in Java 10 ea....

Clearly EFL and elm_code devs do not care about application developers. Why is elm_code in EFL? if you are just going to rely on elm_tests and EDI not real world issues.

Why you all are not testing your changes against ecrire. That says right there, You do not care if EFL/elm_code changes break the ONLY consumer of elm_code other than EDI. That is basically saying API developers do not care to break Application Developers applications....

No one will find that acceptable. All the talk on lists about the website, docs, etc. That is all a total waste of time given the bugs that exist, quality of the code etc. That is time better spent in coding... But I know this will fall on deaf ears as it has. Why EFL is a large active community with tons of complex EFL applications in wide use.... I seriously find it laughable the effort your making to attract EFL application developers. Yet how you all respond and treat said developers. Nothing you do with the website or docs will overcome these technical issues or poor development practices.

Seriously this is a text editor. If elm_code has such issues for a basic text editor. That says more about the abilities of the developers and the quality of their code, or lack there of. Seriously ridiculous! Need to change your stance, approach, and attitude.... Write better code!!!!

Maybe install ecrire, and test out the stuff yourself. You will see most issues come from elm_code/efl. Which I guess is my responsibility to test, etc since I am the one developing elm_code....

Do you really think this sort of mess will attract others to develop applications using EFL? You seriously need to stop and think about that. Stop putting it on others....

**Comment by andydotxyz** [comment]
If you are using a BETA defined API (not declared stable in any release) which you are (see CMakeLists.txt -> "ADD_DEFINITIONS(-DEFL_BETA_API_SUPPORT)" it is requested that you follow it's development. There was a recent discussion about how themes are handled and the usage of some elm_code_widget examples did not work with this change. This applied to all elm widgets, elm_code was one of the widgets that was affected - see this change internally: https://phab.enlightenment.org/rEFL3490da0fffa3aed3109e68d9d26539c446b3c28f
The "legacy style" APIs remain consistent and that is what should have been used all along, I realise now.

It appears that your code followed the same model. The appropriate change to ecrire codebase is (which works with older efl versions as well):

```
-   doc->widget = efl_add(elm_code_widget_class_get(),
-                              doc->win,
-                              elm_obj_code_widget_code_set(efl_added,
-                                                           doc->code));
+   doc->widget = elm_code_widget_add(doc->win, doc->code);
```

If you apply that change you should be able to have it work against efl git and efl releases. And this issue can be closed.

**Comment by wltjr** [comment]
@ajwillia-ms a API you told me to use.... Which you should not have done. You should have just left me alone.... There is also little to no documentation for elm_code. I mentioned that once before. I bet in all your documentation efforts and examples. You have yet to make any docs or examples for elm_code... If it is not ready for such, then you should never tell anyone to develop an application using such... That is not my fault you suggested that. It is my fault for listening and using. I very much regret that and will likely pay for my own mistake moving away from elm_code. Which is what @haxworx recommend and at this point I rather not have to rely on either @ajwillia-ms nor @haxworx given the attitudes and months spent developing something that is still a BETA api... or more like Alpha. Stuff doesn't break like this in beta, and beta means ready for release less unkown issues.

Its clear you all have no concept of proper versioning with major.minor,patch, and clearly do not understand what BETA means.... "Beta phase generally begins when the software is feature complete but likely to contain a number of known or unknown bugs"
https://en.wikipedia.org/wiki/Software_release_life_cycle#Beta

If you are still adding features and making major changes. You are not even alpha, much less beta....

Thank you for suggestion of change, but that again shows your still making major changes. Also since that is not made against any released version. There is not much I can do in build system or otherwise. I would break stuff for people using released versions of EFL to fix an issue in git. Which may change before next release. as you all change up elm_code. You say it will work not, but you also said to use elm_code. Making it hard to trust what you say, since you lead me down the wrong path to begin with.

I would remove elm_code from EFL enitrely. It is clearly not ready to be included. I would also stop telling others to use it. Though likely no one else interested in using elm_code. If I move ecrire back to elm_entry from elm_code. Then EDI will be the only consumer of elm_code and no reason to include it in EFL for just 1 application. No others will have use for such... Clearly not thinking about who will use elm_code. Beyond EDI and ecrire, nothing else will likely use that.

With one of the developers telling me to move back to elm_entry. If I do I will never switch back to elm_code. That makes no sense. Why @haxworx did not start out with the code fix you are suggesting.

What you all are doing here is VERY wrong. How the code is being developed. One person saying use this, another saying not. Making all sorts of noise without looking at code or providing any suggestions on fixes in code. Given you all know the elm_code codebase and actively working on it and changing things... @haxworx  could have made a single comment with the suggestion rather than a bunch of comments that had nothing of technical or any value. Same goes for you @ajwillia-ms so quick to correct me on my tone and attitude etc. Then you realize what's going on. Maybe next time start with the technical fix and leave out all the rest....

For a 1 line correction....

**Comment by wltjr** [comment]
@Peter2121 please test out that commit and see if it resolves the issue for you. I did not see any issues with the change. I guess that is what was causing the issue you were seeing. Though could be others since it seems aspects of the elm_code API is or has changed. Thank you for testing out that change!

**Comment by Peter2121** [comment]
The last commit changes something, but I still have problems with the background color (see the screenshot).
There are some errors in console too (probably, menu-related):
```
$ ecrire
Fontconfig warning: "local.conf", line 1093: saw number, expected matrix
ERR<16727>:edje lib/edje/edje_calc.c:756 _edje_part_description_apply() Cannot find description "visible" in part "base" from group "elm/hover/base/main_menu_submenu/default". Fallback to default description.
ERR<16727>:edje lib/edje/edje_calc.c:756 _edje_part_description_apply() Cannot find description "visible" in part "base" from group "elm/hover/base/main_menu_submenu/default". Fallback to default description.
ERR<16727>:elementary lib/elementary/efl_ui_focus_manager_calc.c:216 node_get() Focusable 0x400000227013 (Elm.Popup) not registered in manager 0x400000007c99

```
![ecrire](https://user-images.githubusercontent.com/3807655/34199350-c7397bba-e56d-11e7-9c81-196ecd81fd0c.jpg)

**Comment by Peter2121** [comment]
It's slightly different (better) with default theme:
![ecrire1](https://user-images.githubusercontent.com/3807655/34200060-0634eee2-e570-11e7-88ef-83a3e533b319.jpg)

**Comment by wltjr** [comment]
That is pretty strange for all that to be white. Do you have any alpha value set? Can you try adjusting that and see if it makes any difference. Not sure what is going on there.

**Comment by wltjr** [comment]
Almost seems like 2 issues, EFL and elm_code. The area where the line numbers and text is is elm_code. The white area seems to be EFL. Like the bottom footer where line/column numbers are located, also the search box. That stuff has nothing to do with elm_code. That seems like general EFL/theme issues or something. I may need to set some background on that stuff that was not required before. Not sure why its showing as white.

**Comment by wltjr** [comment]
Those console errors definitely come from EFL, its missing parts in its default theme and/or they are changing stuff and have not updated the theme. I opened a few bugs along those lines before losing ability to use my phab account. The first one is on your exact error :) Reported in October and remains months later... I can address in my theme, but they need to fix the default stuff...
https://phab.enlightenment.org/T6219
https://phab.enlightenment.org/T6220
https://phab.enlightenment.org/T6225

**Comment by Peter2121** [comment]
I think that have the own theme is not so bad. Or at least fallback to it if something is lost in default one.

**Comment by wltjr** [comment]
I was speaking of my theme [Eminience](https://github.com/Obsidian-StudiosInc/eminence). I can fix those issues in my own theme. I am not sure I could address in any ecrire specific theme. Which I plan to never have a custom theme for ecrire. If I have a theme/edj file for ecrire. it will be for animations, like the animated settings UI in terminology and etui. I plan to make the settings like that someday. If not via code then via edje/edj file. I also may use edje/edj for a desktop or mobile UI layout.

Having my own theme, and having it not replace aspects of the default theme in say etui and verne. I really do not like apps shipping their own theme modifications. If a user changes to a different theme for E those apps do not update properly. They still have aspects of the default theme in their own theme file. Which then to properly create a new theme for all you end up having to create themes for each application.

Like how I have enimence for E/Elementary and one for Terminology. I do not want to have to make one for verne, etui, ecrire and other apps. I really dislike that aspect of edje/edj. It is very hard to inherit stuff and override. You have to literally duplicate to replace. Why I plan to never do that for ecrire. Which means anyone using a custom theme, will see ecrire rendered the way they want via their theme. No parts will keep using the default theme.

Though seems there is some theming now for elm_code, like for the gray line numbers background and black line for the active line. I want to get ride of both of those and just use background colors. I do not like that look. It makes them stand out more than I would like. But that I think I can do in my Eminence theme for elementary :)

**Comment by ApostolosB** [comment]
Don't know if this was an issue a few days back but just compiled and works.

**Comment by wltjr** [comment]
@Peter2121 can you update your EFL from git and see if the problem remains. Thanks!

**Comment by Peter2121** [comment]
The problem remains with the last EFL version, compiled from GIT.

**Comment by wltjr** [comment]
@Peter2121 are you back to the original issue? Did the white area ever go away?

**Comment by Peter2121** [comment]
With the default theme I have the same picture as I posted 20/12. The white area is still here.

**Comment by wltjr** [comment]
Ok thanks good to know. I avoid any EFL/E git stuff as it seems problematic, and not sure what will remain come next release. I will likely wait till closer to or just after release and fix any issues. No clue what is going on I try to follow development. Seems like there should be a 2.x with Wayland, EO, etc but not sure. They are doing it all on 1.x, and its interesting. I cannot even run E 22.1, stuck on E 21.10.

Most the outstanding issues in elm_code that effect ecrire have not been addressed in git. Thus for me little benefit. I think one bug was fixed I could maybe backport to EFL 1.20.6 from git. Otherwise seems like more problems than fixes/solutions. Not sure what that means for ecrire future or other EFL app development. I am not to encouraged sadly. I plan to continue on, but its not a easy path for sure. Seems maybe using elm_code was a big mistake, but then why its in EFL in the first place. I don't see myself going through effort to remove and replace. I would likely look to say pickup Figleaf and add syntax and other missing support to that and maybe switch to GTK dev vs EFL. Given I could fix other issues in like Claws Mail, vs working on say EFL port of Claws Mail or other stuff.

Another chicken egg thing, they want more EFL apps, but not really anything conducive to such from community to code... Very sorry for any issues you run into with Ecrire. I wish I had more ability to fix and seems some issues nothing is happening on. I will have to look into fixing to have any hope of it ever being resolved... Makes me neglect my own open issues on Ecrire and the whole thing is just not good. Though I have mostly been busy updating Java stuff for Java 9, including Netbeans 9. Which I use Netbeans for coding Ecrire and other stuff, Java, C etc. I would never use EDI for all the issues with elm_code. Much of which effect it as well, though some unique issues like this to Ecrire... Fun stuff!!!

**Comment by andydotxyz** [comment]
It seems strange to level this squarely at EFL. The issue cannot be replicated in any Elementary tests or other users of the elm_code API.
Isn't it possible there may be an error in ecrire that could be fixed to solve this?...

**Comment by wltjr** [comment]
If it is an error in Ecrire, it is happening due to changes in EFL. Take that how ever you will. Typically in other situations, if a toolkit changes and breaks applications. They tend to not blame the applications. There are no real users of elm_code other than ecrire and edi. Given the number of other issues open for some time that are confirmed elm_code. Really hard to have confidence that it is not issues there.

Even the last correction came again as some change in EFL required a change in how elm_code was initialized. That is not so much ecrires fault. It was one way and changed. This issue does not exist under EFL 1.20.6, it exists in git due to changes since 1.20.6 was released. Which seem to effect and break ecrire. Rather than upstream carrying their changes broke ecrire. They want ecrire to be fixed. That makes a lot of sense...

Basically every EFL update can break application. Which then applications need to fix themselves to work around EFL changes... It seems you all do not care about anything other than your stuff. Maybe stop using elm_code tests, and rely more on ecrire to test out elm_code and its implementation. Nothing else uses elm_code in any real sense.

It is possible it is the result of some other change in EFL that may not be elm_code related, not sure. Either way Ecrire code has not changed, EFL has, thus the issue.

**Comment by wltjr** [comment]
My plan here is to wait till next release of EFL. See what the status is and address any changes in Ecrire to fix as needed, if needed by then. I do not run EFL from git, and do not really have time to test such.

**Comment by andydotxyz** [comment]
Efl does commit to not breaking published, public apis. Elm_code sadly is still in beta and so is not quite the same. The change to which you refer was due to a move in EFL interfaces which are also still unreleased.

Apologies if there was some confusion here.

**Comment by wltjr** [comment]
No worries I understand. Its why I do not run it from git, as things may change etc. I understand elm_code to be beta as well. Though usually stuff does not change as much in beta. All just reasons to see where things stand come next release. It works with EFL 1.20.6. Anyone choosing to use git risks the issues. Not sure how to make that more known to them, any warnings likely would not be seen.

I would not be to concerned with this at least not till next release. Other issues to address that would be more beneficial. Hopefully what ever changed in EFL is either resolved by next released. Or come time for that I can see about what's changing.

Also I would expect it more for like 1.21 vs 1.20.7. Which if 1.21, then it would make sense to have some issues. I rather not put in conditional code now. Worse case I will make what ever changes and update it build requirements to latest release. Or if very trivial maybe conditionals if there are people still wanting to run older EFL.

**Comment by wltjr** [comment]
Testing EFL 1.20.7-pre I am not seeing any issues. I think this is a git issue, if it still exists. I may modify build system to prevent building against git. Since such issues will not be resolved till EFL release. Which the same blockers should prevent that till its known to work. Not sure if we should keep this open. If it does, I will add a git tag to reflect this issue only exists when using EFL from git.

**Comment by wltjr** [comment]
@Peter2121 I am able to confirm this issue with EFL 1.21_beta1. I will see what I need to do in order to fix. Which will make ecrire require 1.21+. Pretty sure any changes will break it for 1.20. Fun stuff... :face_with_head_bandage:

**Comment by wltjr** [comment]
It is a little better now, the main area of elm_code is no longer white. I assume integers that were unsigned when signed ones were wanted were causing that problem. The bottom bar is still white, so something is still off. It is spitting out a ton of stuff on shutdown and other stuff when running.

**Comment by wltjr** [comment]
The issue seems to come from a change in how transparency is handled in EFL 1.21. It is the alpha setting causing the issues. If you max it out to 100% alpha, no transparency. Then close and open, everything should be a solid color depending on theme. I am working on a solution now that I have narrowed down the problem. It did not exist at all in EFL 1.20. I can make an ecrire window pretty transparent.
![shot-2018-07-25_19-47-03](https://user-images.githubusercontent.com/12835340/43233376-8f5fcab0-9043-11e8-8cbb-23a11ec0c323.jpg)

That is not working in EFL 1.21, aspects end up white. A commit to fix premultiple color error messages fixed everything but the background. Which causes the bottom to still be white. I am working on that, and  hopefully do not have to lose the alpha transparency ability.

**Comment by wltjr** [comment]
In fixing that, the bottom area now has no background on EFL 1.20 or 1.21... I will open a new issue for that.

**Comment by wltjr** [comment]
Turns out this is an upstream issue as mentioned in https://github.com/Obsidian-StudiosInc/ecrire/issues/34#issuecomment-407943202. There are issues with setting the alpha value for elm_bg. It is no longer respected. Thus the window was white.

**Comment by haxworx** [comment]
There are issues.

We've introduced theming support. This is probably causing you problems.

On a lighter note you now have accesss to 4 colour-schemes.

**Comment by wltjr** [comment]
@haxworx this issue is not a elm_code issue. It is a elm_bg issue. Something changed with elm_bg. It no longer respects alpha value, ```evas_object_color_set``` on object created by ```elm_bg_add```. The rest of the objects still respect alpha and become transparent. At the moment the bottom bar of Ecrire is fully transparent as is the window.

I did have to fix other calls to ```evas_object_color_set``` to handled pre-multiplied alpha. f019671. Once I did that it corrected all elements except the background. Which I removed for the time being in 51eb5af4 causing the bottom to be fully transparent. But other elements transparency is working fine with the background gone. If I add it back, it negates the alpha of other objects. That seems to be a regression or new issue in EFL 1.21. Its existed since this was first opened. So some commit around that date.

On another note, I added some stuff to my theme [Eminence for elm_code](https://github.com/Obsidian-StudiosInc/eminence/commit/fff8b04bd58c2faf921f5a049e0a463f6ea17dc0). Which at the time did not work. I think I am noticing that working now in EFL 1.21. Thank you for that!

**Comment by wltjr** [comment]
If you do this, it should be gray and transparent. It ends up being solid gray. The alpha value is not respected it is always solid. never transparent. The alpha value should cause the object to become transparent or solid.
```
elm_win_alpha_set (win, EINA_TRUE);

Evas_Object *bg;
bg = elm_bg_add(win);
evas_object_color_set(bg, 128, 128, 128, 128);
```

That is what caused this issue initially. f019671 was necessary to fix transparency on other items. But the background still ends up solid despite that change. Which is some issue with the elm_bg not respecting alpha anymore in EFL 1.21.

**Comment by wltjr** [comment]
This is what the problem reported in the initial comment looks like
![e-5b61d949eb4750 77893733](https://user-images.githubusercontent.com/12835340/43533766-9f7b9c72-9583-11e8-8386-f0574fafe5ad.jpg)

Switching to pre-multiple alpha fixed most objects in f019671. I had to remove the background in  51eb5af so the other transparent objects can be transparent looks like this.
![e-5b61da2bc8b097 98397528](https://user-images.githubusercontent.com/12835340/43533837-cd59f134-9583-11e8-9fb5-61e115ee5a95.jpg)

But need to be able to do ```evas_object_color_set``` on ```elm_bg``` which presently the alpha portion is ignored only on elm_bg, no other objects. Not sure why.

**Comment by wltjr** [comment]
This is what happens when I set Alpha on elm_bg in 1.21. The first is no alpha the 2nd is full alpha, thus all white/gray. Other elements alpha works but not elm_bg. When other stuff becomes transparent you see more of the white/gray background.

No alpha on elm_bg -> RGBA set to  255
![shot-2018-08-07_13-40-24](https://user-images.githubusercontent.com/12835340/43792710-9272327a-9a47-11e8-83e1-209c83cdd611.jpg)

full alpha on elm_bg -> RGBA set to 128
![shot-2018-08-07_13-40-39](https://user-images.githubusercontent.com/12835340/43792925-147a514e-9a48-11e8-8663-21ff720abd47.jpg)

**Comment by wltjr** [comment]
I am pretty sure commit Enlightenment/efl@25136dd or Enlightenment/efl@81ba035 broke transparency on backgrounds. I have reviewed the [transparency code in edi](https://github.com/Enlightenment/edi/blob/273384d916fa3eeb7d97c735bf4e339306afb6f5/src/bin/edi_theme.c#L20), and doing the same in ecrire. I still end up with the same as above.

The only way to get the color is via the new API
```c
efl_gfx_color_get(efl_part(_win, "background"), &r, &g, &b, &a);
fprintf(stderr, "_bg r %d g %d b %d a %d\n",r, g, b, a);

_bg r 64 g 64 b 64 a 255
```
But soon as I do a ```efl_gfx_color_set``` it goes back to black. Even using pre-multiplied RGB values. Doing ```efl_gfx_color_set(_bg, 64, 64, 64, 255)``` should have no visible change. But if I do that, it becomes black. If I try to do ```efl_gfx_color_set```on the object returned from ```efl_part``` it has no effect. Something seems way off.

This is what ```efl_gfx_color_set(_bg, 64, 64, 64, 255)``` looks like. See the black at the bottom. That should be gray like main menu at the top.
![shot-2018-08-24_23-04-25](https://user-images.githubusercontent.com/12835340/44614416-26391700-a7f2-11e8-8310-78b97c5c3966.jpg)
